trace/doc.go

```diff
@@ -0,0 +1,164 @@
+// interface's documentation for details.
+//
+// Additional information on Spans
+//
+// There are several additional things that can be put on a Span: the
+// most common ones are Errors, Tags and Samples. Spans can also be
+// indicators. The following sections will talk about these in detail.
+//
+// Error handling
+//
+// Call span.Error with the error that caused a Span to
+// fail. Different programs have different requirements for when a
+// Span should be considered to have "failed". A typical rule of thumb
+// is that if you have a traced function that returns an error and it
+// would return an error, that function should also use
+// span.Error. This way, anyone who views a trace visualization will
+// be able to tell that an error happened there. To report an error,
+// use the following:
+//
+//   if err != nil {
+//           span.Error(err)
+//           return err
+//   }
+//
+// Tags
+//
+// Spans can hold a map of name-value pairs (both strings) on their
+// Tags field. These will be reported to trace visualization tools as
+// indexed tags, so it should be possible to search for spans with the
+// relevant information.
+//
+// Note: a Span object, when created, will have a nil Tags map to
+// reduce the number of unnecessary allocations. To add Tags to a
+// Span, it is recommended to initialize the Tags object with a
+// constructed map like so:
+//
+//   span.Tags = map[string]string{
+//           "user_id": userID,
+//           "action": "booped",
+//   }
+//
+// Samples
+//
+// Veneur's trace API supports attaching SSF Samples to Spans. These
+// can be metrics, service checks or events. Veneur will extract them
+// from the span and report them to any configured metrics provider.
+//
+// In the future, veneur may add information that allows linking the
+// extracted metrics to the spans on which they were reported.
+//
+// See the ssf package and the trace/metrics package for details on
+// using Samples effectively.
+//
+// Indicator Spans
+//
+// Typically, a traced service has some units of work that indicate
+// how well a service is working. For example, a hypothetical
+// HTTP-based payments API might have a charge endpoint. The time that
+// an API call takes on that endpoint and its successes or failures
+// can be used to extract service-level indicators (SLI) for latency,
+// throughput and error rates.
+//
+// Concretely, the veneur server privileges these Indicator Spans by
+// doing exactly the above: It reports a timer metric tagged with the
+// service name on the span, and a flag indicating whether the span
+// had Error set.
+//
+// To report an indicator span, use:
+//
+//   span.Indicator = true
+//
+// OpenTracing Compatibility
+//
+// Package trace's data structure implement the OpenTracing
+// interfaces. This allows users to apply all of OpenTracing's helper
+// functions, e.g. to extract a Span from the context, use
+// SpanFromContext and a type assertion:
```

Comment:
Idea: Is it worth adding a "provided for experimental compatibility with opentracing" tag to every method/function here? That would make it clearer to people jumping around through the docs which is which.

If you think so, I can take doing this.

Reply:
Oooh, that'd be great, would make it much easier when we later rip out the parts that relate to OT ((: